sc-memory/sc-memory/tests/sc-memory/units/agents/test_sc_action.cpp

```diff
@@ -182,6 +182,92 @@ TEST_F(ScActionTest, GenerateActionAndSetGetResult)
 }
 
+TEST_F(ScActionTest, GenerateActionAndSetGetArcAsResult)
+{
+  ScAddr const & testClassAddr = m_ctx->GenerateNode(ScType::ConstNodeClass);
+  ScAction action = m_ctx->GenerateAction(testClassAddr);
+
+  EXPECT_THROW(action.GetResult(), utils::ExceptionInvalidState);
+  EXPECT_FALSE(action.InitiateAndWait(1));
+
+  EXPECT_THROW(action.GetResult(), utils::ExceptionInvalidState);
+  ScAddr const & nodeAddr1 = m_ctx->GenerateNode(ScType::ConstNodeStructure);
+  ScAddr const & arcAddr1 = m_ctx->GenerateConnector(ScType::ConstPermPosArc, nodeAddr1, nodeAddr1);
+  EXPECT_NO_THROW(action.SetResult(arcAddr1));
+  EXPECT_NO_THROW(action.SetResult(arcAddr1));
+
+  ScAddr const & nodeAddr2 = m_ctx->GenerateNode(ScType::ConstNodeStructure);
+  ScAddr const & arcAddr2 = m_ctx->GenerateConnector(ScType::ConstPermPosArc, nodeAddr2, nodeAddr2);
+  EXPECT_NO_THROW(action.SetResult(arcAddr2));
+  EXPECT_FALSE(m_ctx->IsElement(arcAddr1));
+  EXPECT_THROW(action.SetResult(arcAddr1), utils::ExceptionInvalidParams);
+  EXPECT_NO_THROW(action.SetResult(arcAddr2));
+
+  action.FinishSuccessfully();
+  EXPECT_THROW(action.SetResult(arcAddr2), utils::ExceptionInvalidState);
+  ScStructure const & result = action.GetResult();
+  EXPECT_TRUE(result.IsEmpty());
+  EXPECT_EQ(result, action.GetResult());
+  EXPECT_EQ(result, arcAddr2);
+}
+
+TEST_F(ScActionTest, GenerateActionAndSetResultIfOtherResultExists)
+{
+  ScAddr const & testClassAddr = m_ctx->GenerateNode(ScType::ConstNodeClass);
+  ScAction action = m_ctx->GenerateAction(testClassAddr);
+
+  EXPECT_THROW(action.GetResult(), utils::ExceptionInvalidState);
+  EXPECT_FALSE(action.InitiateAndWait(1));
+
+  EXPECT_THROW(action.GetResult(), utils::ExceptionInvalidState);
+
+  ScAddr const & structureAddr1 = m_ctx->GenerateNode(ScType::ConstNodeStructure);
+  ScAddr const & arcAddr = m_ctx->GenerateConnector(ScType::ConstCommonArc, action, structureAddr1);
+  m_ctx->GenerateConnector(ScType::ConstPermPosArc, ScKeynodes::nrel_result, arcAddr);
+
+  ScAddr const & structureAddr2 = m_ctx->GenerateNode(ScType::ConstNodeStructure);
+  action.SetResult(structureAddr2);
+
+  EXPECT_THROW(action.FinishSuccessfully(), utils::ExceptionInvalidState);
+}
+
+TEST_F(ScActionTest, GenerateActionAndSetResultIfSameResultExists)
+{
+  ScAddr const & testClassAddr = m_ctx->GenerateNode(ScType::ConstNodeClass);
+  ScAction action = m_ctx->GenerateAction(testClassAddr);
+
+  EXPECT_THROW(action.GetResult(), utils::ExceptionInvalidState);
+  EXPECT_FALSE(action.InitiateAndWait(1));
+
+  EXPECT_THROW(action.GetResult(), utils::ExceptionInvalidState);
+
+  ScAddr const & structureAddr1 = m_ctx->GenerateNode(ScType::ConstNodeStructure);
+  ScAddr const & arcAddr = m_ctx->GenerateConnector(ScType::ConstCommonArc, action, structureAddr1);
+  m_ctx->GenerateConnector(ScType::ConstPermPosArc, ScKeynodes::nrel_result, arcAddr);
+
+  action.SetResult(structureAddr1);
+
+  EXPECT_NO_THROW(action.FinishSuccessfully());
```

Comment:
yeah nice
please add EXPECT_EQ(GetResult(), structureAddr) and it will be excellent